### PR TITLE
feat: add missing schema validation

### PR DESCRIPTION
**Pull request by quesebifurcan**
Validate extracted entities against provided schemas. 

In `aether.kernel.api.views.validate_mappings`,
differentiate between an invalid request (400) and an
unforeseen extraction exception (500).

Add a MappingValidationSerializer.

Add tests for entity validation.

Example submission response for validation errors:
```
{
    "id": "6c6579bb-0367-4b79-9944-63de62b3aab4",
    "url": "http://kernel.aether.local:8000/submissions/6c6579bb-0367-4b79-9944-63de62b3aab4/",
    "mapping_url": "http://kernel.aether.local:8000/mappings/ed0883ea-23f1-4afc-b0a1-8a7e8d21d70d/",
    "entities_url": "http://kernel.aether.local:8000/entities/?submission=6c6579bb-0367-4b79-9944-63de62b3aab4",
    "attachments_url": "http://kernel.aether.local:8000/attachments/?submission=6c6579bb-0367-4b79-9944-63de62b3aab4",
    "attachments": [],
    "created": "2018-05-16T08:38:22.244513Z",
    "modified": "2018-05-16T08:38:22.272596Z",
    "revision": "1",
    "map_revision": "1",
    "date": "2018-05-16T08:38:22.272630Z",
    "payload": {
        "firstname": 1,
        "lastname": "b",
        "aether_errors": [
            {
                "type": "EntityValidationError",
                "description": "Extracted record did not conform to registered schema",
                "data": {
                    "id": "c28e0ed9-c628-4b60-91f9-b3337a807dd8",
                    "firstName": 1,
                    "familyName": "b"
                }
            }
        ],
        "aether_extractor_enrichment": {
            "Person": {
                "id": [
                    "c28e0ed9-c628-4b60-91f9-b3337a807dd8"
                ]
            }
        }
    },
    "mapping": "ed0883ea-23f1-4afc-b0a1-8a7e8d21d70d"
}
```

Resolves: https://jira.ehealthafrica.org/browse/AET-242

### CODE REVIEW

**Comment by obdulia-losantos** [comment]
LGTM!

**Comment by obdulia-losantos** [comment]
Could it be possible to include a better error handling when the mapping does not include an `id` for the used entities? So far is an obscure 400 error.

Not sure if here or in another PR...

**Comment by quesebifurcan** [comment]
@obdulia-losantos Yes, good point, we should fix that. I'd prefer to address that in a separate PR, if that's ok with you?

**Review by obdulia-losantos — DISMISSED** [review]

File: aether-kernel/conf/pip/primary-requirements.txt [hunk 1]
```diff
@@ -22,6 +22,8 @@ drf-dynamic-fields
 drf-openapi
 jsonpath_ng
 jsonschema
+ipdb
+ipython
```
Comment (tooxie) [inline aether-kernel/conf/pip/primary-requirements.txt]: Are these dependencies going to be deployed to production as well?

Comment (quesebifurcan) [reply aether-kernel/conf/pip/primary-requirements.txt]: Good point and yes, at the moment they will be. Perhaps we should keep them out of this file until we have a separate requirements file for development?

File: aether-kernel/aether/kernel/api/views.py [hunk 2]
```diff
@@ -174,27 +174,24 @@ def validate_mappings(request):
...
+            type_errors = submission_data['aether_errors']
+            errors = type_errors or jsonpath_errors
+            return Response({
+                'entities': [entity.payload for entity in entities],
+                'mapping_errors': errors,
+            })
+        except Exception as e:
+            return Response(str(e), status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+    else:
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```
Comment (tooxie) [inline aether-kernel/aether/kernel/api/views.py]: You could return early if the serializer is not valid and save 1 indentation level. Remember: *flat is better than nested* 😋 

Also, it seems to me that there are too many lines in that `try`. Is there any logging done of the errors besides sending them to the client?

File: aether-kernel/conf/pip/primary-requirements.txt (hunk 1 above)
Comment (tooxie) [reply aether-kernel/conf/pip/primary-requirements.txt]: Yes, I would prefer that :)

File: aether-kernel/aether/kernel/api/tests/test_mapping_validation.py [hunk 3]
```diff
@@ -158,16 +158,20 @@ def test_validate_mapping__failure(self):
...
-                error_message=mapping_validation.MESSAGE_NO_MATCH,
-            ),
+            mapping_validation.JsonpathValidationError.from_failure(
+                mapping_validation.Failure(
+                    path=mapping_definition['mapping'][0][1],
+                    description=mapping_validation.MESSAGE_NO_MATCH,
+                ),
+            ).as_dict(),
+            mapping_validation.JsonpathValidationError.from_failure(
+                mapping_validation.Failure(
```
Comment (tooxie) [inline aether-kernel/aether/kernel/api/tests/test_mapping_validation.py]: Why are `path` and `description` wrapped in a `Failure` object just to be then unwrapped by `JsonpathValidationError.from_failure()`? You could directly instantiate it:
```py
JsonpathValidationError(
    path=mapping_definition['mapping'][1][0],
    description=mapping_validation.MESSAGE_NO_MATCH
)
```

File: aether-kernel/aether/kernel/api/mapping_validation.py [hunk 4]
```diff
@@ -23,10 +23,37 @@
 MESSAGE_NO_MATCH = 'No match for path'
 
 Success = collections.namedtuple('Success', ['path', 'result'])
-Failure = collections.namedtuple('Failure', ['path', 'error_message'])
+Failure = collections.namedtuple('Failure', ['path', 'description'])
+
+
+class JsonpathValidationError(object):
```
Comment (tooxie) [inline aether-kernel/aether/kernel/api/mapping_validation.py]: I would have expected it to extend `Exception` given the name. And looking at how it's used, it looks like a function could have been enough:
```py
def failure_to_dict(failure):
    return {
        'description': failure.description,
        'path': failure.path,
        'type': 'JsonpathValidationError',
    }
```

Have you considered that approach?

File: aether-kernel/aether/kernel/api/utils.py [hunk 5]
```diff
@@ -401,7 +416,32 @@ def extract_entity(entity_type, entities, requirements, data, entity_stub):
...
+            schema_definition = schemas[entity_name]
+            try:
+                validate_payload(
+                    schema_definition=schema_definition,
+                    payload=entity_payload,
+                )
+                validated_entities[entity_name].append(entity_payload)
+            except EntityValidationError as err:
+                error = {
+                    'type': EntityValidationError.__name__,
```
Comment (tooxie) [inline aether-kernel/aether/kernel/api/utils.py]: Why not just use a string literal here? The name will always be *'EntityValidationError'*, right?

File: aether-kernel/aether/kernel/api/views.py [hunk 6]
```diff
@@ -174,27 +198,18 @@ def validate_mappings(request):
...
+            'entities': [entity.payload for entity in entities],
+            'mapping_errors': errors,
         })
     except Exception as e:
-        return Response(
-            {'message': 'Entity extraction error'},
-            status=status.HTTP_400_BAD_REQUEST
-        )
+        error = str(e)
+        logger.exception(error)
```
Comment (tooxie) [inline aether-kernel/aether/kernel/api/views.py]: It would be very useful to log the traceback as well as the error message.

**Comment by quesebifurcan** [comment]
Summary of updates:

1. Removed confusing `JsonpathValidationError` class. As Alvaro pointed out, it looked too much like an exception.
2. Include traceback in error log.
3. Change severity level of log message from `exception` to `debug`.
4. Rewrite `validate_setter()` to handle flat destination schemas (following up on today's discussion in #aether-dev).
5. (Finally) allow multiple validation errors (entity validation errors, jsonpath lookup errors) to be returned from the `validate-mappings` endpoint.
6. Removed confusing "type" information in responses (the descriptions provide enough information). It would be nice to have but is not a requirement for now. We can revisit this as part of https://jira.ehealthafrica.org/browse/AET-262.

@tooxie

File: aether-kernel/aether/kernel/api/tests/test_views.py [hunk 7]
```diff
@@ -420,10 +429,22 @@ def test_example_entity_extraction__failure(self):
...
-        self.assertEqual(len(response_data['entities']), 1)
-        self.assertEqual(len(response_data['mapping_errors']), 2)
-
-    def test_example_entity_extraction__assert_raises(self):
+        self.assertEqual(len(response_data['entities']), 0)
+        self.assertEqual(len(response_data['mapping_errors']), 3)
+        expected = set([
+                'Could not find schema "person"',
+                'No match for path',
+                'Extracted record did not conform to registered schema',
```
Comment (tooxie) [inline aether-kernel/aether/kernel/api/tests/test_views.py]: A couple of unimportant nitpicks:
* Why do you turn the lists into sets? You could just compare the lists.
* Extra spacing :)

**Review by tooxie — DISMISSED** [review]
Feel free to ignore my comments, I think the PR is good to go as is.

File: aether-kernel/aether/kernel/api/tests/test_views.py (hunk 7 above)
Comment (quesebifurcan) [reply aether-kernel/aether/kernel/api/tests/test_views.py]: Ah, good catch. The `set()` thing is just habit, really. I'll fix that along with the spacing (why didn't the linter catch that? :confused:)

**Review by tooxie — APPROVED** [review]

**Pull request merged**